apm-agent-core/src/main/java/co/elastic/apm/agent/impl/transaction/TraceContext.java

```diff
@@ -161,17 +188,29 @@ public boolean asChildOf(TraceContext child, Object ignore) {
     };
 
     public static <C> boolean containsTraceContextTextHeaders(C carrier, TextHeaderGetter<C> headerGetter) {
-        return headerGetter.getFirstHeader(TRACE_PARENT_TEXTUAL_HEADER_NAME, carrier) != null;
+        return headerGetter.getFirstHeader(W3C_TRACE_PARENT_TEXTUAL_HEADER_NAME, carrier) != null;
     }
 
     public static <C> void removeTraceContextHeaders(C carrier, HeaderRemover<C> headerRemover) {
-        headerRemover.remove(TRACE_PARENT_TEXTUAL_HEADER_NAME, carrier);
+        headerRemover.remove(W3C_TRACE_PARENT_TEXTUAL_HEADER_NAME, carrier);
+        headerRemover.remove(ELASTIC_TRACE_PARENT_TEXTUAL_HEADER_NAME, carrier);
+        headerRemover.remove(TRACESTATE_HEADER_NAME, carrier);
+        headerRemover.remove(TRACE_PARENT_BINARY_HEADER_NAME, carrier);
     }
 
-    public static <S, D> void copyTextHeaders(S source, TextHeaderGetter<S> headerGetter, D destination, TextHeaderSetter<D> headerSetter) {
-        String elasticApmTraceParent = headerGetter.getFirstHeader(TRACE_PARENT_TEXTUAL_HEADER_NAME, source);
+    public static <S, D> void copyTraceContextTextHeaders(S source, TextHeaderGetter<S> headerGetter, D destination, TextHeaderSetter<D> headerSetter) {
+        String w3cApmTraceParent = headerGetter.getFirstHeader(W3C_TRACE_PARENT_TEXTUAL_HEADER_NAME, source);
+        if (w3cApmTraceParent != null) {
+            headerSetter.setHeader(W3C_TRACE_PARENT_TEXTUAL_HEADER_NAME, w3cApmTraceParent, destination);
+        }
+        String elasticApmTraceParent = headerGetter.getFirstHeader(ELASTIC_TRACE_PARENT_TEXTUAL_HEADER_NAME, source);
         if (elasticApmTraceParent != null) {
-            headerSetter.setHeader(TRACE_PARENT_TEXTUAL_HEADER_NAME, elasticApmTraceParent, destination);
+            headerSetter.setHeader(ELASTIC_TRACE_PARENT_TEXTUAL_HEADER_NAME, elasticApmTraceParent, destination);
+        }
+        // copying only the first tracestate header
+        String tracestate = headerGetter.getFirstHeader(TRACESTATE_HEADER_NAME, source);
```

Comment:
This assumes that the source is guaranteed to only have one `tracestate` header, which might not be true. 

Why not iterate over all values and call `setHeader` for each value?

Reply:
This is what I wrote in the comment above. I don't really want to spend resources on that, it is only used for a very specific scenario at the moment.

> Why not iterate over all values and call setHeader for each value?

It is not conforming to the `tracestate` definition anyway. Let's live that as is